src/algebraic_geometry/height_function.lean

```diff
@@ -0,0 +1,48 @@
+import data.real.basic
+import algebra.category.Group.abelian
+import group_theory.finiteness
+import data.set.finite
+import some_lemmas
+
+open_locale classical
+open finset
+
+universe u
+
+variable (A : Ab.{u})
+
+structure height_function :=
+(to_fun : A → ℝ)
+(nonneg : ∀ P, 0 ≤ to_fun P)
+(C1 : A → ℝ)
+(height_add_le : ∀ (Q P : A), to_fun (P + Q) ≤ 2 * to_fun P + C1 Q)
+(m : ℕ)
+(C2 : ℝ)
+(height_nsmul_ge : ∀ (P : A), ↑(m+2)^2 * to_fun P - C2 ≤ to_fun ((m+2) • P))
+(finite : ∀ (C : ℝ), set.finite { P : A | to_fun P < C })
+
+variables {A} {m : ℕ}
+
+abbreviation M : add_subgroup A :=
+{ carrier := { x | ∃ y, x = m • y },
+  zero_mem' := ⟨0, by rw nsmul_zero⟩,
+  add_mem' := λ x y ⟨x', hx⟩ ⟨y', hy⟩, ⟨x' + y', by rw [nsmul_add, ←hx, ←hy]⟩,
+  neg_mem' := λ x ⟨x', hx⟩, ⟨-x', by rw [neg_nsmul, ←hx]⟩ }
```

Comment:
I think this is just `m • (⊤ : add_subgroup A)`?

Reply:
Could you please tell me where `m • (⊤ : add_subgroup A)` is defined? I want to add a lemma saying that `a ∈ m • N ↔ ∃ a' : N, a = m • a'`

Reply:
`submonoid.pointwise`

Reply:
Thank you.